### ISSUE TITLE
Central plugin classpath management

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#1370]
## Motivation

This is a very rough idea, drawing from the pain points that we have collected from our community, the insights we got from the plugin DSL work last year and some discussion we had on the performance team. Centralizing plugin classpath management would solve many different issues

### Performance and correctness 

- If a plugin is loaded twice in two different projects and these two projects try to exchange some of the plugin's types (e.g. through configuration attribute matching), they will either fail or have to do expensive serialization operations.
- Resolving/Loading the same plugin multiple times in different projects adds an unnecessary performance overhead
- If a class was already loaded by a parent classloader, then the buildscript block in a subproject gives the user a wrong sense of control. Even if she adds a different version of a library to the subproject buildscript classpath, Gradle will still use the one from the parent classloader

### Structuring build logic

The fact that script plugins cannot properly apply other external plugins frustrates many users (#1262). If a script plugin applies another external plugin, the classes of that external plugin are only visible to the script plugin, but not to the project that the plugin is applied to. This was "kind of okay" with dynamic Groovy, but is really bad for statically typed Kotlin scripts.

The same is true for custom Gradle distributions that want to inject and apply plugins across all builds of an organization. Currently this is very awkward and some use cases don't work at all (#1322).

### Consistent DSL

Not all plugins are hosted on the plugin portal, but still we'd like to end up with a consistent DSL in project build scripts. This should be the `plugins {}` DSL, because it is concise and allows us to find out which plugins a project has very early in the lifecycle. This is important for IDE integration.

## Proposal

Instead of specifying the plugin classpath in project build scripts, define it centrally in `settings.gradle`. 

- add a `plugins` block to the upcoming `pluginManagement` DSL (see #1343)
- all plugins that should be available in this build can be specified there
- deprecate `buildscript` block and `version` in the `plugins {}` block for projects
- support script plugins in that `plugins {}` block
- support plugins with custom Maven coordinates (i.e. not hosted on the plugin portal)
- do conflict resolution between all plugins and their dependencies, including dependencies of script plugins

A typical settings.gradle file would look like this:

```
pluginManagement {
  repositories {
    maven { url '...'}
    gradlePluginPortal()
  }
  plugins {
    id 'foo' version '1.0' //from the portal
    id 'bar' by 'org.example:bar:2.1' //from a Maven repo
    id 'baz' from 'gradle/baz.gradle' //a local script plugin 
    id 'fizz' from 'http:\\.../fizz.gradle' //a remote script plugin
  }
}
```

All build scripts would then just refer to those ids and see all the plugins, including their dependencies:

```
plugins {
  id 'foo'
  id 'bar'
}

//cross-configuration also works as expected
subprojects {
  apply plugin: 'baz'
  apply plugin: 'fizz'

  tasks.withType(BuzzTask) { //transitively brought in by the 'fizz'  script plugin
   //type-safe configuration
  }
}
```

## Potential follow-ups

Once we have central management for the plugin classpath, we are in a much better position to handle other long-standing issues.

### Plugin implementation isolation

Allow script plugins to separate their api and implementation dependencies instead of having just a "classpath". Only do conflict resolution on the api dependencies and have an isolated classloader for the implementation. 

Do the same for binary plugins, based on their published metadata (e.g. for Maven only expose the `compile` dependencies in the shared classloader, but hide the `runtime` dependencies in an isolated classloader).

Next, allow plugins (both script and binary) to define which of their classes are part of their API and only expose those to build scripts. This allows plugins to evolve more easily because they don't need to worry about users calling their internals.

**Comment by oehme** [comment]
@eriwen @adammurdoch I just needed to get this off my mind, nothing to act upon right now. But any feedback is welcome :)

**Comment by JLLeitschuh** [comment]
Would you also be able to configure the `buildSrc` plugins via this method?

It would be nice to be able to centralize configuring repository mirrors for companies that use mirroring technology. Having to reuse the same logic three places (your buildSrc project, your normal build `buildscript` blocks and your normal `repositories` configuration for resolving java dependencies).

**Comment by oehme** [comment]
I don't think so, because `buildSrc` is another, isolated build, similar to a composite build. In fact, we plan to replace `buildSrc` with composite builds completely. 

You can avoid most of the duplication though by writing a small script that configures your company mirror both as plugin repository and as the repository for all project dependencies. Then you would apply that script at the top of every build's `settings.gradle` file. `buildSrc` would be nothing special in that regard, it would have its own `settings.gradle` file and be essentially a standalone build.

**Comment by remal** [comment]
Don't close this issue. This is an automatic message by [Fresh](https://github.com/c-hive/fresh) - a bot against stale bots.

**Comment by martinbonnin** [comment]
> Allow script plugins to separate their api and implementation dependencies instead of having just a "classpath". Only do conflict resolution on the api dependencies and have an isolated classloader for the implementation.

One nice thing with this is that tasks wouldn't be necessarily invalidated whenever the buildscript changes. The build cache, etc.. could stay valid even though a new plugin is added.

**Comment by rkrisztian** [comment]
> Allow script plugins to separate their api and implementation dependencies instead of having just a "classpath". Only do conflict resolution on the api dependencies and have an isolated classloader for the implementation.

I so need this! It happened to me so many times that I wanted to upgrade a single plugin, but I had to fight other plugins about the preferred version of some low-level transitive dependency like `org.apache.commons:commons-lang3`.

**Comment by martinbonnin** [comment]
I'll allow myself a small shameless plug for [Gratatouille](https://github.com/GradleUp/gratatouille/), a framework for building Gradle plugins in Kotlin that comes with support for [classloader isolation](https://github.com/GradleUp/gratatouille/?tab=readme-ov-file#classloader-isolation-optional) and separating plugin api (wiring) and implementation (tasks). 

I've been working on it for the past year or so and it's now used in [nmcp](https://github.com/GradleUp/nmcp) and [Apollo](https://github.com/apollographql/apollo-kotlin) to the point where I'm starting to feel confident about it. 

It's no replacement for a first party API like described here but if it can help mitigate the issue in the meantine, I'd welcome all feedbacks!
